src/main/java/com/google/cloud/discotoproto3converter/proto3/DocumentToProtoConverter.java

```diff
@@ -91,14 +96,44 @@ private void readSchema(Document document) {
   private void resolveReferences(Message message) {
     for (Field field : message.getFields()) {
       Message valueType = field.getValueType();
-      if (valueType.isRef()) {
+      if (valueType.isRef()) { // replace the field object with a link to the message it references
         field.setValueType(protoFile.getMessages().get(valueType.getName()));
       } else {
         resolveReferences(valueType);
       }
     }
   }
 
+  private boolean checkForAllowedAnyFields(Message message) {
+    return checkForAllowedAnyFields(message, message.getName());
+  }
+
+  private boolean checkForAllowedAnyFields(Message message, String previousFieldPath) {
+    // We want to we recursively check every child and don't short-circuit when haveAny becomes
+    // true, as we rely on the side effect (exception) to signal a google.protobuf.Any in an
+    // unsupported location.
+    boolean haveAny = false;
+    for (Field field : message.getFields()) {
+      Message valueType = field.getValueType();
+      String currentFieldPath = previousFieldPath + "." + field.getName();
+      if (valueType.getName() == Message.PRIMITIVES.get("google.protobuf.Any").getName()) {
+        if (currentFieldPath.endsWith(".error.details")) {
+          haveAny = true;
+          if (trace) {
+            System.err.printf("Found ANY field at %s\n", currentFieldPath);
+          }
+        } else {
+          throw new IllegalArgumentException(
+              "illegal ANY type not under \"*.error.details\": " + currentFieldPath);
+        }
+      } else {
+        // order matters here: we want to not ensure the first check happens
```

Comment:
nit: I would rewrite this with an `if` statement to avoid `||` and its unintended short circuiting effect, just for the sake of readability.

Reply:
Well, not an `if`, but did make the side effect more explicit.